proposals/type-classes.md

```diff
@@ -237,60 +202,155 @@ fun addInts(a: Int, b: Int, with Monoid<Int>): Int = add(a, b)
 fun addInts(a: Int, b: Int): Int = with(intext.IntMonoid) { add(a, b) }
 ```
 
-## Compile resolution rules
+## Type Class Instance Rules
 
-When the compiler finds a call site invoking a function that has type class instances constrains declared with `with` as in the example below:
+Classical interfaces only allow the implementation of interfaces to occur when a type is defined. Type classes typically relax this rule and allow implementations outside of the type definition. When relaxinng this rule it is important to preserve the coherency we take for granted with classical interfaces.
 
-Declaration: 
+For those reasons type class instances must be declared in one of two places:
+
+1. In the same file as the type class definition (interface-side implementation)
```

Comment:
Would it be ok to be in the same package instead of file? Users and Lib authors may want to organize their instances in separate files.

Reply:
How would a user define instances for types they have not authored?. For example someone creating library `C` that depends on `Arrow` and the Kotlin `stdlib` wants to define a `Foldable` instance for a collection type in the `stdlib` that Arrow has chosen not to provide instances for alongside the definition of `Foldable`.

Reply:
What you're referring to are type class orphans. If you read the section on orphan instances you can see the difficulty of correctly handling orphan instances without causing surprising and incorrect handling. Somebody creating library `C` should **not** be allowed to define a type class instance for a type and type class they don't control; that library would then be incompatible with any library that defined the same instance.

Imagine if adding a library to your dependencies could silently change how the `Array` interface works for some types.

Relaxing it to "same package" allows users to define such orphan instances by using that package directive. Packages on the JVM are really just conventions. This is the reason that Kotlin requires "same file" for sealed classes.

Reply:
The issue there and now speaking in terms of libraries like Arrow is that FP is not backed into Kotlin and we can't foresee all the data types that our type classes may support, nor we can make other libs provide instances for our type classes. That would make integration libraries that attempt to fill voids or lack of FP support in other Kotlin libs impossible. As an example a generic derivation lib may want to provide generic instance derivation for product types such as data classes defined in the std lib like `Pair`. `Generic` may be a type class in Arrow and Arrow may choose to not provide instances for `Pair` because it already has it's own `Tuple2` isomorphic representation. 

As an author of a library I may want to expose instances that are optimized and if these instances obey the type class laws declared by the original type class author I may want to be able to provide such instances for third party consumers. Would it be ok if a compilation error is raised when two conflicting instances are found and make the user disambiguate via internal override which one he wants to use?:
```kotlin
// arrow is in the class path and defines Foldable and an instance of Foldable<List>
// the stdLib defines List and no instances
// third party library X provides instance for Foldable<List>
fun <F, A> doWithFoldable(fa: Kind<F, A>, with: Foldable<F>) = ...

doWithFoldable(listOf(1)) // fails to compile because there are two instances and the user shall disambiguate
```
```kotlin
internal extension object Foldable<List> : XLibListFoldable

doWithFoldable(listOf(1)) // compiles because the user has explicitly selected which instance should be use
```

Reply:
This is not unique to Kotlin or type-classes, it is a problem with all interface mechanisms in all programming languages. And those languages (including Kotlin) have existing mechanisms to work around that, generally either extending or wrapping the type. Both of those are applicable here.

```kotlin
// classical interfaces
import foo.collections.FooList
import bar.collections.Foldable

class MyFooList : FooList(), Foldable {
    ...
}
// or
class MyFooList(private val underlying: FooList) : Foldable {
    ...
}

// type classes
import foo.collections.FooList
import bar.collections.Foldable

class MyFooList : FooList() {
    ...
}
extension object : Foldable<MyFooList> {
    ...
}
// or
class MyFooList(private val underlying: FooList) {
    ...
}
extension object : Foldable<MyFooList> {
    ...
}
```
Both of these approaches have the benefit that you can't accidentally mix incompatible implementations of the interface. In the multiple instances approach the compiler can't track that the instances are being threaded correctly.

It's generally polite when writing libraries with data types to implement all applicable interfaces from the stdlib. Likewise with type-classes it is polite when writing a library with a type-class to provide implementations for all relevant types in the stdlib. In the Rust and Haskell communities people have generally been good about this and accept pull requests to add new instances.

I'd like to give Kotlin library authors the benefit of the doubt rather than trying to preemptively route around uncooperative library authors with language features.

Reply:
To go more into your example:
```kotlin
// arrow is in the class path and defines Foldable and an instance of Foldable<List>
// the stdLib defines List and no instances
// third party library X provides instance for Foldable<List>
fun <F, A> doWithFoldable(fa: Kind<F, A>, with: Foldable<F>) = ...
```
How is the compiler going to know that there is an instance in library X? Nothing in our file indicates a dependency on that library. We can no longer look just at a file to determine its full dependencies, we need to have a global instance resolution mechanism. But also for safety it is _vital_ that we do warn here so that the user can thread everything correctly and not mix incompatible instances.

It's possible that this is ultimately the behavior that we want, but it complicates version 1 a lot when we could ship with a simpler version to begin with and add this later if it becomes obvious we need it.

Reply:
I think this would be a great time to mention that we'll definitely need a delegate equivalent for type classes, e.g.
```kotlin
data class Distance(val toInt)
object : Order<Distance> by toInt
```